contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,1283 @@
+use cosmwasm_std::{
+    to_binary, Api, Binary, Deps, DepsMut, Env, HumanAddr, MessageInfo, Order, Response, StdError,
+    StdResult, Uint128, KV,
+};
+use cw_storage_plus::Bound;
+
+use cw0::{maybe_canonical, Event};
+use cw1155::{
+    ApproveAllEvent, ApprovedForAllResponse, BalanceResponse, BatchBalanceResponse,
+    Cw1155BatchReceiveMsg, Cw1155HandleMsg, Cw1155QueryMsg, Cw1155ReceiveMsg, Expiration,
+    IsApprovedForAllResponse, TokenId, TokenInfoResponse, TokensResponse, TransferEvent,
+};
+use cw2::set_contract_version;
+
+use crate::error::ContractError;
+use crate::msg::InitMsg;
+use crate::state::{APPROVES, BALANCES, MINTER, TOKENS};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw1155-base";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const DEFAULT_LIMIT: u32 = 10;
+const MAX_LIMIT: u32 = 30;
+
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InitMsg,
+) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let minter = deps.api.canonical_address(&msg.minter)?;
+    MINTER.save(deps.storage, &minter)?;
+    Ok(Response::default())
+}
+
+/// To mitigate clippy::too_many_arguments warning
+pub struct ExecuteEnv<'a> {
+    deps: DepsMut<'a>,
+    env: Env,
+    info: MessageInfo,
+}
+
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw1155HandleMsg,
+) -> Result<Response, ContractError> {
+    let env = ExecuteEnv { deps, env, info };
+    match msg {
+        Cw1155HandleMsg::SendFrom {
+            from,
+            to,
+            token_id,
+            value,
+            msg,
+        } => execute_send_from(env, from, to, token_id, value, msg),
+        Cw1155HandleMsg::BatchSendFrom {
+            from,
+            to,
+            batch,
+            msg,
+        } => execute_batch_send_from(env, from, to, batch, msg),
+        Cw1155HandleMsg::Mint {
+            to,
+            token_id,
+            value,
+            msg,
+        } => execute_mint(env, to, token_id, value, msg),
+        Cw1155HandleMsg::BatchMint { to, batch, msg } => execute_batch_mint(env, to, batch, msg),
+        Cw1155HandleMsg::Burn {
+            from,
+            token_id,
+            value,
+        } => execute_burn(env, from, token_id, value),
+        Cw1155HandleMsg::BatchBurn { from, batch } => execute_batch_burn(env, from, batch),
+        Cw1155HandleMsg::ApproveAll { operator, expires } => {
+            execute_approve_all(env, operator, expires)
+        }
+        Cw1155HandleMsg::RevokeAll { operator } => execute_revoke_all(env, operator),
+    }
+}
+
+fn checked_add(a: Uint128, b: Uint128) -> StdResult<Uint128> {
```

Comment:
Interesting. We define `sub` for `Uint128` to possibly return an error if it underflows (not enough balance). We don't do this for adding as we assumed this would never happen. 

We do however, [test that this extremely rare condition causes a panic](https://github.com/CosmWasm/cosmwasm/blob/main/packages/std/src/math.rs#L774-L780), which is enough to keep the contract secure. It won't provide a pretty error message to the user, but will block the attack vector. I would just use the normal `Uint128 + Uint128` approach

Reply:
I reverted back to simple `+` operator.
Opened issue in cosmwasm-std for the overflow handling: https://github.com/CosmWasm/cosmwasm/issues/850